paddle/operators/roi_pool_op.cu

```diff
@@ -0,0 +1,265 @@
+        int c = (index / pooled_width / pooled_height) % channels;
+        int n = index / pooled_width / pooled_height / channels;
+
+        const int64_t* offset_input_rois = input_rois + n * 5;
+        int roi_batch_ind = offset_input_rois[0];
+        int roi_start_w = round(offset_input_rois[1] * spatial_scale);
+        int roi_start_h = round(offset_input_rois[2] * spatial_scale);
+        int roi_end_w = round(offset_input_rois[3] * spatial_scale);
+        int roi_end_h = round(offset_input_rois[4] * spatial_scale);
+
+        int roi_width = max(roi_end_w - roi_start_w + 1, 1);
+        int roi_height = max(roi_end_h - roi_start_h + 1, 1);
+        T bin_size_h = static_cast<T>(roi_height)
+                      / static_cast<T>(pooled_height);
+        T bin_size_w = static_cast<T>(roi_width)
+                      / static_cast<T>(pooled_width);
+
+        int hstart = static_cast<int>(floor(static_cast<T>(ph) * bin_size_h));
+        int wstart = static_cast<int>(floor(static_cast<T>(pw) * bin_size_w));
+        int hend = static_cast<int>(ceil(static_cast<T>(ph + 1) * bin_size_h));
+        int wend = static_cast<int>(ceil(static_cast<T>(pw + 1) * bin_size_w));
+
+        hstart = min(max(hstart + roi_start_h, 0), height);
+        hend = min(max(hend + roi_start_h, 0), height);
+        wstart = min(max(wstart + roi_start_w, 0), width);
+        wend = min(max(wend + roi_start_w, 0), width);
+        bool is_empty = (hend <= hstart) || (wend <= wstart);
+
+        T maxval = is_empty ? 0 : -FLT_MAX;
+        int maxidx = -1;
+        const T* offset_input_data =
+            input_data + (roi_batch_ind * channels + c) * height * width;
+        for (int h = hstart; h < hend; ++h) {
+          for (int w = wstart; w < wend; ++w) {
+            int input_data_index = h * width + w;
+            if (offset_input_data[input_data_index] > maxval) {
+              maxval = offset_input_data[input_data_index];
+              maxidx = input_data_index;
+            }
+          }
+        }
+        output_data[index] = maxval;
+        if (argmax_data) {
+          argmax_data[index] = maxidx;
+        }
+    }
+  }
+
+template <typename T>
+__global__ void GPURoiPoolBackward(
+    const int nthreads,
+    const int64_t* input_rois,
+    const T* output_grad,
+    const int64_t* argmax_data,
+    const int num_rois,
+    const float spatial_scale,
+    const int channels,
+    const int height,
+    const int width,
+    const int pooled_height,
+    const int pooled_width,
+    T* input_grad) {
+    int index = blockIdx.x * blockDim.x + threadIdx.x;
+    int offset = blockDim.x * gridDim.x;
+    for (int i = index; i < nthreads; i += offset) {
+      int pw = index % pooled_width;
+      int ph = (index / pooled_width) % pooled_height;
+      int c = (index / pooled_width / pooled_height) % channels;
+      int n = index / pooled_width / pooled_height / channels;
+
+      const int64_t* offset_input_rois = input_rois + n * 5;
```

Comment:
How about add a comment about the magic number 5, or define it a constexpr value?

Reply:
done